pytidb/fusion.py

```diff
@@ -132,3 +134,103 @@ def fusion_result_rows_by_rrf(
+    vs_method: DistanceMetric,
+    fts_rows: List[Row],
+    get_row_key: Callable[[Row], RowKeyType],
+    vs_weight: Optional[float] = 0.5,
+    fts_weight: Optional[float] = 0.5,
+) -> Tuple[List[str], List[Row]]:
+    """
+    Fusion the search results by weights.
+
+    Args:
+        vs_rows: Vector search list of result rows
+        fts_rows: Full text search list of result rows
+        get_row_key: Function to get the key (primary key or _tidb_rowid) from a row
+        vs_weight: The weight of the vector search list. Must be a number between 0 and 1. Default is 0.5.
+        fts_weight: The weight of the full text search list. Must be a number between 0 and 1. Default is 0.5.
+    Returns:
+        Tuple containing:
+        - List of field names
+        - List of fused result rows sorted by RRF score
+    """
+    if not vs_rows and not fts_rows:
+        return [], []
+
+    if vs_weight <= 0 or vs_weight >= 1 or fts_weight <= 0 or fts_weight >= 1:
+        raise ValueError("weight must be a number between 0 and 1")
+
+    # Calculate weighted scores for each result in both lists
+    weighted_scores = {}
+
+    # Process first list
+    for i, row in enumerate(vs_rows):
+        vs_distance = row._mapping["_distance"]
+
+        if vs_method == DistanceMetric.COSINE:
+            normalized_vs_distance = _normalize_score(vs_distance, "cosine")
+        elif vs_method == DistanceMetric.L2:
+            normalized_vs_distance = _normalize_score(vs_distance, "l2")
+        else:
+            raise ValueError("Invalid distance metric")
+
+        key = get_row_key(row)
+        weighted_scores[key] = normalized_vs_distance * vs_weight
+
+    # Process second list and add scores
+    for i, row in enumerate(fts_rows):
+        row_score = row._mapping["_match_score"]
+        normalized_row_score = _normalize_score(row_score, "full_text")
+        key = get_row_key(row)
+        if key in weighted_scores:
+            weighted_scores[key] += normalized_row_score * fts_weight
+        else:
+            weighted_scores[key] = normalized_row_score * fts_weight
+
+    # Merge rows.
+    fusion_strategies = {
+        "_score": lambda a, b, row_a, row_b, key: weighted_scores[key],
+    }
+    all_fields, merged_rows = merge_result_rows(
+        vs_rows, fts_rows, get_row_key, fusion_strategies
+    )
+
+    # Sort rows by weighted score.
+    sorted_rows = sorted(
+        merged_rows, key=lambda row: row._mapping["_score"] or 0, reverse=True
+    )
+
+    return all_fields, sorted_rows
+
+def _normalize_score(original_value: float, score_type: Literal["l2", "cosine", "full_text"]) -> float:
+    """Normalize the score to a value between 0 and 1.
+
+    Args:
+        original_value: The score to normalize
+        score_type: The type of score to normalize
+
+    Returns:
+        The normalized score
+    """
+
+    if score_type == "cosine":
+        # The input value is the cosine distance, the original range is [-1, 1].
```

Comment:
The range of cosine distance is [0, 2]?

Reply:
Nope, it is $$[-1, 1]$$. FYI: https://en.wikipedia.org/wiki/Cosine_similarity
Unless we did something special.

<img width="647" alt="image" src="https://github.com/user-attachments/assets/61b5fd14-48cb-4594-9a16-270728fb16c0" />

Reply:
The range of cosine similarity is [-1, +1], but cosine distance's range is [0, 2]

Reply:
Got it, let me correct it then.